payloadinterceptor.go

```diff
@@ -15,42 +15,67 @@
 package otelconnect
 
 import (
+	"context"
+	"errors"
+	"io"
 	"sync"
+	"time"
 
 	connect "connectrpc.com/connect"
 )
 
 type streamingClientInterceptor struct {
 	connect.StreamingClientConn
 
-	receive func(any, connect.StreamingClientConn) error
-	send    func(any, connect.StreamingClientConn) error
-	onClose func()
+	ctx     context.Context //nolint:containedctx
+	state   *state
+	startAt time.Time
 
+	requestStarted sync.Once
 	mu             sync.Mutex
 	requestClosed  bool
 	responseClosed bool
-	onCloseCalled  bool
+}
+
+func (s *streamingClientInterceptor) init() {
+	s.requestStarted.Do(func() {
+		header := s.StreamingClientConn.RequestHeader()
+		s.ctx = s.state.start(s.ctx, header)
+	})
 }
 
 func (s *streamingClientInterceptor) Receive(msg any) error {
-	return s.receive(msg, s.StreamingClientConn)
+	s.init()
+	if err := s.StreamingClientConn.Receive(msg); err != nil {
+		if !errors.Is(err, io.EOF) {
+			s.state.setError(err)
+		}
+		return err
+	}
+	s.state.receive(s.ctx, msg)
+	return nil
 }
 
 func (s *streamingClientInterceptor) Send(msg any) error {
-	return s.send(msg, s.StreamingClientConn)
+	s.init()
```

Comment:
This could trivially result in deadlock: this calls `RequestHeader`, which _blocks_ until the request headers are received. However, the server may choose to not send headers until after it has received and processed data. So we must allow sending messages without blocking for request headers.

Reply:
Shoot, never mind. Ignore this comment 🤦.

I was apparently seeing "ResponseHeader" even though this actually says "RequestHeader". My bad. Pay this no mind, and I'll continue reviewing.